keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1096 @@
+
+Using label selectors might be relevant in case of [Story 1](#story-1), as
+MultiKueue can have many jobs in the management cluster. Keeping the jobs in
+the Job controller's informer, and starting the `syncJob` function for them is
+wasteful.
+
+However, the label selector semantics does not support OR-ing these two cases.
+For this reason, we need to ensure all jobs have the label set. One mechanism
+to consider is to use the `defaults.go` file to set the label. However, this
+might have a drawback, that it will require many writes to etcd after a cluster
+upgrade. This concern can be addressed, to some extent, by incubating the Beta
+version which sets the default for new jobs, for a release or more, before
+using `defaults.go`. Then, it seems reasonable to assume that only a small
+portion of jobs would require an actual etcd write after cluster upgrade.
+
+We add evaluation of the strategy to [GA requirements](#ga).
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable by passing `controllers=-job,*` in the manifest for of
+`kube-controller-manager`.
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
+)
+```
+
+#### Implementation overview
+
+In the [Beta](#beta) iteration of the feature we skip synchronization of the
+Jobs with the `managed-by` label, if it has any different value than `job-controller`.
+We also default the label for all newly created jobs, to avoid implicit defaults.
```

Comment:
I'm not sure about *requiring* the label on all jobs. I was expecting the *normal* default controller to keep working as-is for jobs without the label, and to switch to excluding jobs with the label, without needing to decorate or modify *any* existing jobs. That would mean that only exceptional jobs that wanted to be managed by a non-default controller would opt into the label.

> We also default the label for all newly created jobs

Does this mean 1) we'd set a label on a job created by a user, or 2) our controllers that create jobs (e.g. cronjob controller) would populate a label? If 1, is there precedence for that? What component would default that and how would it know what job controller name the "default" controller was running as?

Reply:
I don't think we want to change the fact that we're using shared-informers. If that is true, then simple "if skip" is the only path forward.